src/accelerate/accelerator.py

```diff
@@ -359,6 +386,19 @@ def prepare(self, *args):
             - `torch.nn.Module`: PyTorch Module
             - `torch.optim.Optimizer`: PyTorch Optimizer
         """
+        if self.distributed_type == DistributedType.FSDP:
+            model_count = 0
+            optimizer_present = False
+            for obj in args:
+                if isinstance(obj, torch.nn.Module):
+                    model_count += 1
+                if isinstance(obj, torch.optim.Optimizer):
+                    optimizer_present = True
+            if model_count > 1 and optimizer_present:
+                raise ValueError(
+                    "For FSDP to work, prepare must be called for all the models before optimizers are created"
+                )
```

Comment:
Is this still necessary?

Reply:
Yes, currently we will only be able to handle the case of single model and hence throwing error in case of multiple models.

Reply:
The error message could be made clearer then.

Reply:
Done.